be/src/formats/parquet/level_builder.cpp

```diff
@@ -0,0 +1,656 @@
+                                                 const ::parquet::schema::NodePtr& node, const ColumnPtr& col,
+                                                 const CallbackFunction& write_leaf_callback) {
+    const auto data_col = get_raw_data_column<TYPE_BOOLEAN>(col);
+    const auto null_col = get_raw_null_column(col);
+
+    // Use the rep_levels in the context from caller since node is primitive.
+    auto rep_levels = ctx._rep_levels;
+    auto def_levels = _make_def_levelsV3(ctx, node, null_col);
+
+    // sizeof(bool) depends on implementation, thus we cast values to ensure correctness
+    auto values = new bool[col->size()];
+    DeferOp defer([&] { delete[] values; });
+
+    int offset = 0;
+    for (size_t i = 0; i < col->size(); i++) {
+        if (null_col != nullptr && null_col[i] == 0) {
+            values[offset++] = static_cast<bool>(data_col[i]);
+        }
+    }
+
+    write_leaf_callback(LevelBuilderResult{
+            .num_levels = ctx.num_levels(),
+            .def_levels = def_levels->data(),
+            .rep_levels = rep_levels->data(),
+            .values = reinterpret_cast<uint8_t*>(values),
+            .null_bitset = nullptr,
+    });
+
+    return Status::OK();
+}
+
+template <LogicalType lt, ::parquet::Type::type pt>
+Status LevelBuilder::_write_int_column_chunk(const LevelBuilderContext& ctx, const TypeDescriptor& type_desc,
+                                             const ::parquet::schema::NodePtr& node, const ColumnPtr& col,
+                                             const CallbackFunction& write_leaf_callback) {
+    const auto data_col = get_raw_data_column<lt>(col);
+    const auto null_col = get_raw_null_column(col);
+
+    // Use the rep_levels in the context from caller since node is primitive.
+    auto rep_levels = ctx._rep_levels;
+    auto def_levels = _make_def_levelsV3(ctx, node, null_col);
+
+    using source_type = RunTimeCppType<lt>;
+    using target_type = typename ::parquet::type_traits<pt>::value_type;
+
+    if constexpr (std::is_same_v<source_type, target_type>) {
+        // Zero-copy for identical source/target types
+        // If leaf column has null entries, provide a bitset to denote not-null entries.
+        write_leaf_callback(LevelBuilderResult{
+                .num_levels = ctx.num_levels(),
+                .def_levels = def_levels ? def_levels->data() : nullptr,
+                .rep_levels = rep_levels ? rep_levels->data() : nullptr,
+                .values = reinterpret_cast<uint8_t*>(data_col),
+                // Make bitset to denote not-null values
+                .null_bitset = col->has_null() ? _make_null_bitset(col->size(), null_col).data() : nullptr,
+        });
+    } else {
+        // If two types are different, we have to cast values anyway
+        std::vector<target_type> values;
+        values.reserve(col->size());
+        for (size_t i = 0; i < col->size(); i++) {
+            if (null_col == nullptr || null_col[i] == 0) {
```

Comment:
```
if (null_col == nullptr) {
  // use memset here
} else {
  // use masking and memset for fast population, refer to `ArrayColumn::filter_range`
}
```

Reply:
Avoid adding branch in any for-loop

Reply:
We cannot directly manipulate memory since source and target types are different. 

One way to eliminate if-branch in for-loop is to 
1. upgrade column (e.g. int16 -> int32) to match physical type
2. pass level builder result as in the zero-copy scenario

This could lead to much waste when the column is very sparse

Reply:
Updated every write function of primitive types to enable auto vectorization. 

> We cannot directly manipulate memory since source and target types are different.
> 
> One way to eliminate if-branch in for-loop is to
> 
> 1. upgrade column (e.g. int16 -> int32) to match physical type
> 2. pass level builder result as in the zero-copy scenario
> 
> This could lead to much waste when the column is very sparse